src/directLineStreaming.ts

```diff
@@ -0,0 +1,345 @@
+  private disconnectionHandler(e: any) {
+    if (this.connectionStatus$.value === ConnectionStatus.Connecting ||
+        this.connectionStatus$.value === ConnectionStatus.Ended) {
+      return;
+    }
+
+    if (this.token === null) {
+      this.activitySubscriber.error("Token unavailable");
+      return;
+    }
+
+    this.connectionStatus$.next(ConnectionStatus.Connecting);
+    this.retryCount--;
+    if (this.retryCount > 0) {
+      setTimeout(this.streamingWebSocketActivity$.bind(this), this.getRetryDelay());
+    } else {
+      console.warn("Exhausted retries");
+      this.activitySubscriber.error(e);
+    }
+  }
+
+  private streamingWebSocketActivity$(): Observable<Activity> {
+    if (this.activitySubscriber) {
+      this.theStreamHandler.setSubscriber(this.activitySubscriber);
+      this.connectAsync().then(_ => _);
+    } else {
+      return Observable.create(async (subscriber: Subscriber<Activity>) => {
+        this.activitySubscriber = subscriber;
+        this.theStreamHandler = new StreamHandler(subscriber, this.connectionStatus$);
+        await this.connectAsync();
+      });
+    }
+  }
+
+  private async waitUntilOnline() {
+    return new Promise<void>((resolve, reject) => {
+      this.connectionStatus$.subscribe((cs) => {
+        if (cs === ConnectionStatus.Online) return resolve();
+      },
+        (e) => reject(e));
+    })
+  }
+
+  private async connectAsync() {
+    let re = new RegExp('^http(s?)');
+    if (!re.test(this.domain)) throw ("Domain must begin with http or https");
+    let urlSearchParams = new URLSearchParams({token: this.token, conversationId: this.conversationId}).toString();
+    let wsUrl = `${this.domain.replace(re, 'ws$1')}/conversations/connect?${urlSearchParams}`;
+    try {
+      this.streamConnection = new BFSE.WebSocketClient({
+        url: wsUrl,
+        requestHandler: this.theStreamHandler,
+        disconnectionHandler: this.disconnectionHandler.bind(this)
+      });
+
+      await this.streamConnection.connect();
+      let request = BFSE.StreamingRequest.create('POST', '/v3/directline/conversations');
+      let response = await this.streamConnection.send(request);
+      if (response.statusCode !== 200) throw new Error("Connection response code " + response.statusCode);
+      if (response.streams.length !== 1) throw new Error("Expected 1 stream but got " + response.streams.length);
+      let responseString = await response.streams[0].readAsString();
+      let conversation = JSON.parse(responseString);
+      this.conversationId = conversation.conversationId;
+      this.parentDirectLine.referenceGrammarId = conversation.referenceGrammarId;
+      this.connectionStatus$.next(ConnectionStatus.Online);
+
+      // Wait until DL consumers have had a chance to be notified
+      // of the connection status change.
+      await this.waitUntilOnline();
+      this.theStreamHandler.flush();
+      this.retryCount = MAX_RETRY_COUNT;
+    } catch (e) {
+      console.warn(e);
```

Comment:
You may need to set `connectionStatus$` to error or something. Otherwise, it won't reconnect.

Reply:
The reconnect logic does the right thing. Setting it to something else seems to break webchat.

Reply:
If Web Chat is broken, file a bug to Web Chat.